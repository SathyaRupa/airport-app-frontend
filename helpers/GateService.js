import axios from 'axios';
import url from './Url';

export default {
  fetchAll: async page => {
    try {
      const response = await axios.get(`${url}/gates?page=${page}`);
      return response.data;
    } catch (error) {
      console.log('Error fetching data: ', error);
    }
  },
};
