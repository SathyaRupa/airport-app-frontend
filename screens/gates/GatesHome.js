import ItemCard from '../../components/ItemCard';
import * as React from 'react';
import {useEffect, useState} from 'react';
import GateService from '../../helpers/GateService';
import {FlatList, ActivityIndicator} from 'react-native';
import {Avatar} from 'react-native-paper';
import {useIsFocused} from '@react-navigation/native';

function GatesHome({navigation}) {
  const [allDataLoaded, setAllDataLoaded] = useState(false);
  const [fetchedPages, setFetchedPages] = useState([]);
  const [gates, setGates] = useState([]);
  const [loading, setLoading] = useState(true);
  const [page, setPage] = useState(0);
  const isFocused = useIsFocused();

  useEffect(() => {
    async function fetchData() {
      if (!allDataLoaded && isFocused && !fetchedPages.includes(page)) {
        const response = await GateService.fetchAll(page);
        setFetchedPages(prevPages => [...prevPages, page]);
        if (response.length === 0) {
          setAllDataLoaded(true);
        } else {
          setGates(prevGates => [...prevGates, ...response]);
        }
      }
      setLoading(false);
    }
    fetchData();
  }, [page, allDataLoaded, isFocused]);

  const loadMoreGates = () => {
    if (!allDataLoaded) {
      setPage(prevPage => prevPage + 1);
    }
  };

  useEffect(() => {
    if (isFocused) {
      setGates([]);
      setPage(0);
      setAllDataLoaded(false);
      setLoading(true);
      setFetchedPages([]);
    }
  }, [isFocused]);

  const renderFooter = () => {
    return loading ? <ActivityIndicator size="large" color="#4D869C" /> : null;
  };

  const handlePress = id => {
    navigation.push('Gate Details', {id});
  };

  const icon = (
    <Avatar.Image
      style={{backgroundColor: 'transparent'}}
      size={50}
      source={require('../../assets/icons/gates.png')}
    />
  );

  return (
    <FlatList
      data={gates}
      renderItem={itemData => (
        <ItemCard
          id={itemData.item.id}
          value={'Gate ' + itemData.item.gate_number}
          icon={icon}
          testId={`item-card-${itemData.index}`}
          onPress={() => handlePress(itemData.item.id)}
        />
      )}
      onEndReached={loadMoreGates}
      onEndReachedThreshold={0.5}
      ListFooterComponent={renderFooter}
    />
  );
}

export default GatesHome;
