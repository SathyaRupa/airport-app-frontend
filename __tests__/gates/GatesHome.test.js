import React from 'react';
import {render, waitFor} from '@testing-library/react-native';
import GateService from '../../helpers/GateService';
import {useIsFocused} from '@react-navigation/native';
import GatesHome from '../../screens/gates/GatesHome';

jest.mock('react-native-linear-gradient', () => 'LinearGradient');
jest.mock('@react-navigation/native', () => ({
  useIsFocused: jest.fn(),
}));
jest.mock('../../helpers/GateService', () => ({
  fetchAll: jest.fn(),
}));

jest.mock('react-native-toast-message', () => 'Toast');
jest.mock('../../components/ToastMessage', () => ({
  SuccessToast: jest.fn(),
  ErrorToast: jest.fn(),
}));

describe('Gates home - get all Gates', () => {
  let mockNavigation;

  beforeEach(() => {
    mockNavigation = {
      push: jest.fn(),
    };
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  it('snapshot test', () => {
    const gatesHome = render(<GatesHome navigation={mockNavigation} />);

    expect(gatesHome).toMatchSnapshot();
  });

  it('should render an item card', async () => {
    useIsFocused.mockReturnValue(true);

    GateService.fetchAll.mockResolvedValueOnce([
      {floor_number: '10', gate_number: '17'},
    ]);
    const gatesHome = render(<GatesHome navigation={mockNavigation} />);
    const itemCard = gatesHome.queryAllByTestId('item-card-0');

    expect(itemCard).toBeTruthy();
    await waitFor(() => {
      expect(GateService.fetchAll).toHaveBeenCalledWith(0);
      expect(gatesHome.getByText('Gate 17')).toBeTruthy();
    });
  });

  it('should not render an item card when no data is recieved', async () => {
    GateService.fetchAll.mockResolvedValue([]);
    const gatesHome = render(<GatesHome navigation={mockNavigation} />);
    expect(gatesHome.queryByTestId('item-card-0')).toBeNull();
    await waitFor(() =>
      expect(gatesHome.queryByText('Jet Airways')).toBeNull(),
    );
  });
});
