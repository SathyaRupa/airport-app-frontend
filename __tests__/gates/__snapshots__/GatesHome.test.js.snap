// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gates home - get all Gates snapshot test 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={[]}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.5}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <ActivityIndicator
        color="#4D869C"
        size="large"
      />
    </View>
  </View>
</RCTScrollView>
`;
